### ISSUE TITLE
Excalidraw as a component

### ISSUE THREAD

**Issue by vjeux** [excalidraw/excalidraw#1542]
There's been two PR already that attempt to make an Excalidraw component (#958, #1493), sorry I have not been responsive. Before we proceed further I'd like to ask a few questions to @mizchi and @ad1992 (and all the people that +1 the issues). We should have done this in the first place before spending a lot of time with the implementation.

Once we are clear about what the use cases are, the API we want, what to do with options and how to implement it. We'll be able to do the work once and get it merged.

1) What are the **concrete** use cases?

I can think that it would be useful to embed images in a wiki. I would like to know exactly what use case you have in mind so we can figure out the best trade-offs for it. Please be as specific as possible.

2) What is the API that you expect from the component?

```js
<Excalidraw ???? />
```

and in particular:
- How do you expect dimensions to be taken into account? Does it 
- How do you expect the scene to be passed? Is it a controlled component, or do you get an callback when it changes, how do you set the original one?
- How do you get the result back? Do you get a json, or a svg, or a png? Multiple ones? (so you can keep the source and you save the svg/png for consumption)

3) What happens with all the options?

- Load/save file, do we need it?
- Collaboration, does it even make sense?
- Export dialog, needed?
- Github banner, should be gone right?
- Zoom in/out, seems useful.
- Languages, do we want to use the language from the container instead?
- All the drawing options, do we want them? Or do we just want black and white.
- Does it need to support mobile? How do we manage scrolling with our own scroll?

4) What implementation do we want to go to?

At the outset, there seems to be two big possibilities: inject the component directly in the DOM, and pray there will be no style conflicts. Or use an iframe.

Based on the answers of all the questions above, we can figure out which is the best trade-off.

Sorry again about all the wasted work so far. Let's go through the proper process and merge it in. I believe that this is a good feature to have.

**Comment by dai-shi** [comment]
> 1. What are the concrete use cases?

Although I don't have a strong opinion how to implement it, I have a concrete use case in my project: https://github.com/dai-shi/remote-faces
Screenshot: https://twitter.com/dai_shi/status/1255429309835157504
For this project, collaboration is important. The current implementation is based on`<iframe>`.

At the first thought, what I would like to do next would be 1) set user name, 2) set background color, 3) warn deleting the content before closing the iframe if it's the last user. There should be many more as we proceed.

My use case could be pretty different from others, so this is just my two cents.

**Comment by ad1992** [comment]
> There's been two PR already that attempt to make an Excalidraw component (#958, #1493), sorry I have not been responsive. Before we proceed further I'd like to ask a few questions to @mizchi and @ad1992 (and all the people that +1 the issues). We should have done this in the first place before spending a lot of time with the implementation.
> 
Three PR's actually till now including https://github.com/excalidraw/excalidraw/pull/1455. Since the prev PR for npm was out of sync so I implemented from scratch but it doesn't have all changes in this since we initially decided to push in batches so I m maintaining it in forked for the usage


> Once we are clear about what the use cases are, the API we want, what to do with options and how to implement it. We'll be able to do the work once and get it merged.
> 
> 1. What are the **concrete** use cases?

There are multiple use cases when you want to customize the Excalidraw from outside
Below are some scenarios
* Always  showing zen mode
* Whenever user is drawing do something additional on parent App for example (user 1 is drawing etc)
* Customize the shapes which need to be shown in excalidraw by passing shapes array
* use some other backend if needed when the scene updates by passing callback
* Pass a class name to the top wrapper if you want to add some additional css
Like this, there can be many more which can be implemented once the wrapper is ready

> 
> I can think that it would be useful to embed images in a wiki. I would like to know exactly what use case you have in mind so we can figure out the best trade-offs for it. Please be as specific as possible.
> 
> 1. What is the API that you expect from the component?
> 
> ```js
> <Excalidraw ???? />
> ```
Yeah something line 
```
<Excalidraw 
  zenModeEnabled
  shapes={Array of shapes}
  onChange={this.onChange}
  className="my-excalidraw"

```
> 
> and in particular:
> 
> * How do you expect dimensions to be taken into account? Does it

This depends, So we should expect the width and height as props and in component did update we should update it as its controlled component now.
If not passed then it takes window inner height and width as done today. So this can be kept as an option and left on the parent to decide. However, we need to take into consideration the parent top and left in all calculations to render excalidraw correctly. I have made all these changes in a separate fork as we decided to push in batches as per previous discussions.


> * How do you expect the scene to be passed? Is it a controlled component, or do you get an callback when it changes, how do you set the original one?

We should call the callback prop if passed so that parent App can do the additonal things as needed

> * How do you get the result back? Do you get a json, or a svg, or a png? Multiple ones? (so you can keep the source and you save the svg/png for consumption)

I would prefer just pass the result in same format we push to server which is JSON. Future we can think if need additional formats.

> 
> 1. What happens with all the options?
> 
> * Load/save file, do we need it?

this we can show, later prop can be added if needed.

> * Collaboration, does it even make sense?

Yes, collaboration I am using but I m using different backend for it so I had to make changes for it to work so this can be solved with the callback.

I am not sure existing excalidraw collaboration is needed since here its getting rendered inside parent App because most probably if people use it they would want to push it their backend.

> * Export dialog, needed?

We can show it but again can be made optional

> * Github banner, should be gone right?

same as above
> * Zoom in/out, seems useful.

Yes this we should keep
> * Languages, do we want to use the language from the container instead?

Yes preferably the language can be a prop

> * All the drawing options, do we want them? Or do we just want black and white.

Again we can keep it for now and later depending upon requests we can make it optional

> * Does it need to support mobile? How do we manage scrolling with our own scroll?

The initial launch can be without mobile I think. I think will have to check what issues we face when rendering with mobile provider and then figure out the solution.

> 1. What implementation do we want to go to?
> 

I think its best to be exported as an npm package
1. The integration becomes easier
2. Its done in react-y way
3. Controlled/Uncontrolled both scenarios are taken care of 

there are two major changes which needs to be done for this
1. Scope the css
2. Add the width height top and left calculations in the main App code.
But none of these will in any way impact the excalidraw App for sure

> At the outset, there seems to be two big possibilities: inject the component directly in the DOM, and pray there will be no style conflicts. Or use an iframe.
> 

Style conflicts I have faced so scoping is needed for injecting in DOM. Iframe can also work for all scenarios but implementation and maintenance wise package will be easier. 
Good amount of custom code just for iframe will have to written for iframe in main APP and in the consumer APP as well for post message which can be exposed as a react api later with extra work
I had initially tried with both approaches exporting as a package is what I would prefer in terms of integration, effort, maintenance and adding more features as well.

> Based on the answers of all the questions above, we can figure out which is the best trade-off.
> 
> Sorry again about all the wasted work so far. Let's go through the proper process and merge it in. I believe that this is a good feature to have.

**Comment by ad1992** [comment]
@vjeux no need to be sorry :) I understand this is a big decision to make. Thanks a lot for your response and initiating the discussion on the proposal :)

**Comment by dwelle** [comment]
@ad1992 most of what you talk about is not exactly a use case. What we need are specific examples of what apps will be built with excalidraw, and what requirements they'll have. 

Importantly, we need to gather real-world use cases, not hypothesized --- because we developers are good at coming up with any number of potential use cases that will end up not important or used at all.

We also need to figure out if we can start with a small proof-of-concept/MVP implementation to help us gauge *real-world* interest, without us having to fully commit and allocate resources into something which end up requiring only a simple, more limited implementation.

I'll notify some of the people who liked the original PR:

@fsubal, @sadnessOjisan, @meganetaaan, @ygkn, @jay-es, @locol23, @qrg, @RoyMcCrain, @yuumi3, @hashrock, @brodybits, @mbj36, @haikyuu, @Martingmb

**Comment by ad1992** [comment]
@dwelle not sure what you meant by the use cases I mentioned are not exactly a use case. I have been working on integrating excalidraw in a product which is an interview-based platform where excalidraw will be used as a whiteboard. The use cases which I mentioned are actually genuine use cases and excalidraw is an awesome app that interviewing based platforms would especially like to integrate.

**Comment by dwelle** [comment]
@ad1992 Ok, I thought you were mainly describing an API. But your requirements seem overly broad. Shouldn't one given app take a more opinionated stance on how it will integrate excalidraw? Is your app going to allow its users to completely customize the UI of its whiteboard component?

**Comment by ad1992** [comment]
> @ad1992 Ok, I thought you were mainly describing an API. But your requirements seem overly broad. 

@dwelle I did describe what I thought the API would look like so that excalidraw can be easily integrated with any App. 
>Shouldn't one given app take a more opinionated stance on how it will integrate excalidraw? 
Sorry I didn't get what you meant here!

What I meant in the API is that we can keep some parts as optional if needed later so that the consumer APP can pass the prop as false if they don't want to show for example the load/save file. But that we can do later not in the initial launch

Is your app going to allow its users to completely customize the UI of its whiteboard component?
No no product will be using zenmode itself of excalidraw. the customizations are not controlled by the users. So I can give you one more use case
Since we use zen mode so by default it will always be collaborative so username is also something I am passing as a prop to excalidraw so that it shows the name along with cursor and this user name is something which can be changed in parent component so this is a usecase for controlled component.

If you think something I mentioned is not a real-world scenario we can iterate over it.

**Comment by ad1992** [comment]
Sorry for troubling... any update on this @vjeux @dwelle ? 
We can iterate over the API of excalidraw If the proposal of exposing it as a component looks fine to the team.

**Comment by vjeux** [comment]
If I understand correctly, your use case is that you are building an interview platform and you want people to be able to collaborate with a shared whiteboard. This means that you are very interested in live collaboration but not that much in terms of changing colors & co.

So based on this, you just need the drawing options at the top and not the side bar, nor exporting options. You want to be able to set/get the collaboration link so that the multiple work together. And I assume you may want to pre-fill the canvas with some value and maybe get the drawing when the interview is over.

That's a specific use case where we can then figure out what kind of options you want out of it.

**Comment by ad1992** [comment]
@vjeux to give you more context I  am using zen mode, collaboration but not with socketio as we want the data to be saved to the backend so triggering callback here during broadcast will actually help, colors and username as well so as to show our color code and username and username can be changed by the user itself.

Prefilling is not needed as of now since we save the data to backend.
So the API is like
```
<Excalidraw 
   width={width}
   height={height}
   zenModeEnabled 
   user={{username:'name;, color: {colors}} 
   onBroadcast={this.onBroadCast}
/>
```

**Comment by vjeux** [comment]
I'm curious, what do you mean by "not with socketio". Do you re-implement all of the synchronization primitives and do you run your own real time backend?

**Comment by ad1992** [comment]
@vjeux we are using firebase for collaboration. We sync the data to firebase whenever it gets broadcasted and similarly pull the data when any change in firebase. 
That is why for updating the scene from a remote collab I had to move the `updateScene` method out of `initializeSocketClient` so that it can be reused.
But this is a very specific use case and not generic one so this can be discussed later if we want to add support for any external backend.

**Comment by stefanli** [comment]
Hey folks. I have another specific use case that I thought I'd chime in with. I'm creating a flash card/note application and I'd like to allow users to embed excalidrawings into my editor. The way I imagine this to work is similar to how "Insert drawing" works in Google Docs or similar apps: 

1. The user clicks "Insert drawing" 
2. A modal pops up with the excalidraw UI.
3. User creates a drawing and hits "save".
4. An inline-image with the drawing result is shown in the editor.
5. User can click edit at which point the excalidraw modal pops up again pre-populated with the existing drawing.

If the decision is to move ahead with this at some point I'm happy to help out with contributions. My initial high level take on what is needed is inline with what was suggested by @ad1992 above with the addition of being able to access the canvas or somehow convert the contents of the excalidraw component to an image.

**Comment by ad1992** [comment]
@stefanli If I get you right, You basically need a callback that gets triggered when user hits the save button in excalidraw and sends the excalidraw elements as an image?
Prepopulatinng  excalidraw with existinng drawinng can be taken care with `initialData`.

**Comment by pA1nD** [comment]
I believe an excalidraw component would be incredible useful but also very tricky to realize.
We're completely hooked with exclidraw and are using it all over the place so we are looking currently into using it in a number of tools. I'm trying to outline it here one use case:

### Internal Project Management Tool
We have an internal project management tool where we track projects and work on issues.
To specify requirements, we create markdown files (markdown editor). Our customer has an access where he/she can review the requirements (rendered markdown).

In the same way as we use markdown to express ourselves, we started to use excalidraw drawings. Currently as exported SVGs. But everytime we need to change something, we need to open the file in excalidraw, change something, export, upload again. To implify we are planning to include excalidraw as an alternative to the markdown editor. So that we can edit it, and the customers can check it out.

Further down the road, we might even end up sketching solutions out with our customer in a realtime session.

### DX ideas / requirements

#### Frontend
```
<Excalidraw 
  data={project.data} 
  editable={true} 
  session={`${EXCALIDRAW_URL}#room=${project.id},${createHash()}`} 
  user={user.name}
  options={{language: 'en', showLoad: false, showSave: false, showSession: false}}
/>
```

| Property | Default value | Description |
| -- | -- | -- |
| data | null | Set the default data |
| editable | true | Allow a user to edit elements and show / hide the edit-related controlls |
| session | null | URL with room to connect session. Valid value: https://penguin.academy/excalidraw#room=106e6b881deb86c5968f,jxL8E8nDQZU34TbigREJpw |
| user | null | The user's username to be displayed in a shared session |

#### Backend

I would imagine to include excalidraw in a similar fashion as e.g. socket.io. So it would allow me simply to include it in an existing server or to create a new one if needed.

**Using with Node http server**
```
const server = require('http').createServer(...)
const excalidraw = require('excalidraw-server')(server)
```

**Using with Express**
As I'm using express, I could include this in my application like this:
```
// server.js
const excalidraw = require('./excalidraw')
app.use('/excalidraw', excalidraw) // I can handle additional authentication here as needed
```

```
// excalidraw.js
const app = require('express')()
const server = require('http').Server(app)
const excalidraw = require('excalidraw-server')(server);

excalidraw.on(...) // expose here a bunch of useful events.
```
 
I am currently not quite sure how the flow with the events would look like, but my most important requirement would be to store the current version in a database (e.g. every 5 seconds or so). Would that mean that I need to have the whole json in memory on the server or is there a smarter way?

**Comment by ad1992** [comment]
@pA1nD I am working on exporting it as a component [here](https://github.com/excalidraw/excalidraw-embed). There are some pending things left and then will release the first version of it very soon. Then we can discuss what more features can be added to it.

**Comment by pA1nD** [comment]
A little offtopic, but we are currently in the very early stage of creating an open source (video) call tool (WebRTC based). The same interface as above would make sense... and it would work perfectly together.

I imagine a chat would also work in a similar fashion.
I am wondering if there are more tools that follow such a API design - and / or if that is an emerging pattern.

**Comment by ad1992** [comment]
> How would the styling work if excalidraw is directly embedded?
You will have to import the `excalidraw.min.css` file in your component itself for styling. 
it will be scoped so won't affect anything outside excalidraw app.

Does that answer your question?

**Comment by pA1nD** [comment]
Yes! That's perfect.

**Comment by ad1992** [comment]
@pA1nD and everyone interested, you can try out https://github.com/excalidraw/excalidraw-embed :)

**Comment by pA1nD** [comment]
@ad1992 Awesome! Tested it out and it looks great :)

2 Questions:

1) It's currently using the excalidraw backend. Are there plans / is there already work going on to release a server package as well?
2) I think it would be awesome to have additional config options (think hide / show / overwrite buttons. E.g. the save button.) and `editable={false}`.

**Comment by ad1992** [comment]
> @ad1992 Awesome! Tested it out and it looks great :)
> 
> 2 Questions:
> 
> 1. It's currently using the excalidraw backend. Are there plans / is there already work going on to release a server package as well?

@pA1nD there is currently no working going on this. If I understand it right you want callback for events like `client-broadcast`, `first-in-room` ?
Could you explain more o how exposing the server would help?

> 2. I think it would be awesome to have additional config options (think hide / show / overwrite buttons. E.g. the save button.) and `editable={false}`. Would that be a good idea to build that into App.js?

Do you mind creating two separate issues for `options` like hide/show buttons and `editable` in 
 https://github.com/excalidraw/excalidraw-embed ? And we can discuss the use case there then.

**Comment by pA1nD** [comment]
1) This is my szenario: I am including the excalidraw in an app where I handle projects, users and various documents. So when they go to a certain sub-page of a project, I'd want them to be on a excalidraw.

First of all, I'd like to have the socket.io server included in my application server so I don't need to run a second infrastructure (unless I decide to do so). This would e.g allow me to handle authentication on express app level.

There are two core requirements:
- If multiple people open the page, they should "automatically" be connected (join the same session). 
- I want to store the latest state in  a database so whenever its closed and opened, I have the latest state available.

I'm not sure how I would be implementing that right now.

**Comment by ad1992** [comment]
> This is my scenario: I am including the excalidraw in an app where I handle projects, users and various documents. So when they go to a certain sub-page of a project, I'd want them to be on a excalidraw.
> 
> First of all, I'd like to have the socket.io server included in my application server so I don't need to run a second infrastructure (unless I decide to do so). This would e.g allow me to handle authentication on express app level.
> 
> There are two core requirements:
> 
> * If multiple people open the page, they should "automatically" be connected (join the same session).

Does the user share a collaborative URL? How do you decide which user is part of which session?

> * I want to store the latest state in a database so whenever it's closed and opened, I have the latest state available. 
If you can save the session URL as well in your DB and pass that session URL to the excalidraw App so. it initializes with that session then won't that help ?

we [onChange[(https://github.com/excalidraw/excalidraw-embed#onChange) callback so won't you be able to save the data to your backend with say some id and restore it when the app again opens?
> 
> I'm not sure how I would be implementing that right now.

**Comment by ad1992** [comment]
closing this as now we have published [@excalidraw/excalidraw](https://www.npmjs.com/package/@excalidraw/excalidraw) 🎉 . Please feel free to open a new issue for new features/improvements.
